firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -235,50 +230,81 @@ private void persistFid(String fid) throws FirebaseInstallationsException {
   }
 
   /**
-   * Registers the FID with FIS servers if FID is in UNREGISTERED state.
+   * Registers the FID with FIS servers if FID is in UNREGISTERED state. Also, refreshes auth token
+   * if expired.
    *
    * <p>Updates FID registration status to PENDING to avoid multiple network calls to FIS Servers.
    */
-  private Task<String> registerFidIfNecessary(
-      PersistedFidEntry persistedFidEntry, AwaitListener listener) {
-    String fid = persistedFidEntry.getFirebaseInstallationId();
-
-    // Check if the fid is unregistered
-    if (persistedFidEntry.getRegistrationStatus() == RegistrationStatus.UNREGISTERED) {
-      updatePersistedFidWithPendingStatus(fid);
-      executeFidRegistration(persistedFidEntry, listener);
-    } else {
-      updateAwaitListenerIfRegisteredFid(persistedFidEntry, listener);
+  private Task<String> registerFidIfNecessary(AwaitListener listener)
+      throws FirebaseInstallationsException {
+    Boolean isNetworkCallRequired = false;
+
+    synchronized (persistedFid) {
+      PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+      if (persistedFidEntry.isNotGenerated()) {
+        throw new FirebaseInstallationsException(
+            "Local storage has no persisted Fid entry.",
+            FirebaseInstallationsException.Status.CLIENT_ERROR);
+      }
+
+      // If FID registration status is REGISTERED and auth token is expired, or registration status
+      // is UNREGISTERED network call iks required.
+      if ((persistedFidEntry.isRegistered() && isAuthTokenExpired(persistedFidEntry))
+          || persistedFidEntry.isUnregistered()) {
+        isNetworkCallRequired = true;
+
+        // Update FID registration status to PENDING to avoid multiple network calls to FIS Servers.
+        persistedFid.insertOrUpdatePersistedFidEntry(
+            persistedFidEntry
+                .toBuilder()
+                .setRegistrationStatus(RegistrationStatus.PENDING)
+                .build());
+      }
     }
 
-    return Tasks.forResult(fid);
-  }
+    PersistedFidEntry updatedPersistedFidEntry = persistedFid.readPersistedFidEntryValue();
 
-  private void updateAwaitListenerIfRegisteredFid(
-      PersistedFidEntry persistedFidEntry, AwaitListener listener) {
-    if (listener != null
-        && persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTERED) {
-      listener.onSuccess();
+    if (updatedPersistedFidEntry.isNotGenerated()) {
+      throw new FirebaseInstallationsException(
+          "Local storage has no persisted Fid entry.",
+          FirebaseInstallationsException.Status.CLIENT_ERROR);
```

Comment:
is this check just making sure we were able to write to disk? it seems redundant with the one above

Reply:
Done.